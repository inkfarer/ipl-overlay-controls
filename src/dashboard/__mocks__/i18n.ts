import i18next from 'i18next';
import { config } from '@vue/test-utils';

jest.mock('i18next-vue', () => ({
    __esModule: true,
    useTranslation: () => ({ t: i18next.t })
}));

beforeAll(async () => {
    await i18next.init({
        lng: 'cimode',
        appendNamespaceToCIMode: true
    });
    config.global.mocks = {
        $t: i18next.t
    };
});
