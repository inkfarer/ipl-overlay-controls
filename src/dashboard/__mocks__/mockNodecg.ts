import Mock = jest.Mock;

export type ReplicantChangeHandler = (newValue?: unknown, oldValue?: unknown) => void;
export type MockReplicant = { value: unknown };
type OnFunction = (event: string, handler: ReplicantChangeHandler) => void;

export class MockNodecg {
    replicants: {[key: string]: MockReplicant};
    listeners: {[key: string]: ReplicantChangeHandler};
    sendMessage: Mock

    constructor() {
        this.replicants = {};
        this.listeners = {};
        this.sendMessage = jest.fn();
    }

    init(): void {
        const self = this;
        window.nodecg = {
            // @ts-ignore: Just a mock
            Replicant(name: string) {
                const replicantValue: { value: unknown, on: OnFunction } = {
                    value: undefined,
                    on: (event: string, handler: ReplicantChangeHandler) => {
                        self.listeners[name] = handler;
                    }
                };
                self.replicants[name] = replicantValue;
                return replicantValue;
            },
            sendMessage: self.sendMessage,
            // @ts-ignore: Just a mock
            log: {
                info: jest.fn()
            }
        };

        // @ts-ignore: It's a mock.
        window.NodeCG = {
            waitForReplicants: jest.fn().mockResolvedValue(true)
        };
    }
}
