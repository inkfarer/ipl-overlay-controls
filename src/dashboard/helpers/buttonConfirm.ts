/*
buttonConfirm

After importing this script to a dashboard panel, add the attribute 'data-requires-confirmation' to any button.
The button will now ask for confirmation when it is first clicked.
When it is clicked a second time, a 'confirm' event is sent by it, which can be listened for by scripts.

Example:

HTML:
    <button id="dangerous-action-button" data-requires-confirmation>DO A THING</button>
JS:
    document.getElementById('dangerous-action-button').addEventListener('confirm', () => {
        // do something
    });
*/

window.addEventListener('DOMContentLoaded', () => {
    function addListener(btn: HTMLButtonElement) {
        if (btn.dataset.hasConfirmListener !== 'true') {
            btn.addEventListener('click', e => {
                const target = e.target as HTMLButtonElement;

                function resetButton() {
                    target.dataset.isClicked = 'false';
                    target.innerText = target.dataset.originalText;
                    target.style.backgroundColor = '';
                }

                if (target.dataset.isClicked === 'true') {
                    target.dispatchEvent(new Event('confirm'));
                    resetButton();
                } else {
                    target.dataset.isClicked = 'true';
                    target.dataset.originalText = target.innerText;
                    target.style.backgroundColor = 'var(--red)';
                    target.innerText = 'are you sure?';

                    setTimeout(() => {
                        resetButton();
                    }, 5000);
                }
            });

            btn.dataset.hasConfirmListener = 'true';
        }
    }

    const buttons = document.querySelectorAll('button[data-requires-confirmation]');
    buttons.forEach(addListener);

    const observer = new MutationObserver(mutations => {
        mutations.forEach(mutation => {
            mutation.addedNodes.forEach(node => {
                if (node.nodeType === Node.ELEMENT_NODE
                    && node.nodeName === 'BUTTON'
                    && (node as HTMLElement).dataset.requiresConfirmation) {

                    const btn = node as HTMLButtonElement;
                    addListener(btn);
                }
            });
        });
    });

    observer.observe(document.body, {
        subtree: true,
        childList: true
    });
});
