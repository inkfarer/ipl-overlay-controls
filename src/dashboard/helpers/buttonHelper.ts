import { colors } from '../styles/colors';

export function setToggleButtonDisabled(
    toggleTrue: HTMLButtonElement,
    toggleFalse: HTMLButtonElement,
    state: boolean
): void {
    toggleTrue.disabled = state;
    toggleFalse.disabled = !state;
}

export function addChangeReminder(
    elements: HTMLElement[] | NodeListOf<HTMLElement>,
    updateButton: HTMLButtonElement
): void {
    elements.forEach(element => {
        if (!element.tagName) return;

        let event;
        if (element.tagName.toLowerCase() === 'input') event = 'input';
        else if (element.tagName.toLowerCase() === 'select') event = 'change';
        else return;

        element.addEventListener(event, e => {
            const target = e.target as HTMLElement;
            target.dataset.edited = 'true';
            updateButton.style.backgroundColor = colors.red;
        });
    });

    updateButton.addEventListener('click', () => {
        elements.forEach(elem => {
            elem.dataset.edited = 'false';
        });
        updateButton.style.backgroundColor = colors.blue;
    });
}
