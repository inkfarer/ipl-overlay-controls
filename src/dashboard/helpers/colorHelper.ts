export function getContrastingTextColor(backgroundColor: string, lightColor = 'white', darkColor = '#333'): string {
    const { r, g, b } = hexToRbg(backgroundColor);
    const yiq = ((r * 299) + (g * 587) + (b * 114)) / 1000;
    return (yiq >= 128) ? darkColor : lightColor;
}

export function hexToRbg(color: string): { r: number, g: number, b: number } {
    color = color.replace('#', '');
    return {
        r: parseInt(color.substr(0, 2), 16),
        g: parseInt(color.substr(2, 2), 16),
        b: parseInt(color.substr(4, 2), 16)
    };
}
