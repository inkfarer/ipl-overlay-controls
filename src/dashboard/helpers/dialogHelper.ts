import { NodecgDialog } from '../types/dialog';

export function closeDialog(dialogName: string): void {
    (nodecg.getDialog(dialogName) as NodecgDialog).close();
}
