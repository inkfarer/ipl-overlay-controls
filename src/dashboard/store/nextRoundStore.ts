import { NextRound } from 'schemas';
import { createStore, Store, useStore } from 'vuex';
import cloneDeep from 'lodash/cloneDeep';
import { InjectionKey } from 'vue';
import { SetRoundRequest } from 'types/messages/rounds';

const nextRound = nodecg.Replicant<NextRound>('nextRound');

export const nextRoundReps = [ nextRound ];

export interface NextRoundStore {
    nextRound: NextRound;
}

export const nextRoundStore = createStore<NextRoundStore>({
    state: {
        nextRound: null
    },
    mutations: {
        setState(store, { name, val }: { name: string, val: unknown }): void {
            this.state[name] = cloneDeep(val);
        },
        setShowOnStream(store, newValue: boolean) {
            nextRound.value.showOnStream = newValue;
        }
    },
    actions: {
        beginNextMatch(): void {
            nodecg.sendMessage('beginNextMatch');
        },
        setNextRound(store, {
            teamAId,
            teamBId,
            roundId
        }: { teamAId: string, teamBId: string, roundId: string }): void {
            nodecg.sendMessage('setNextRound', { teamAId, teamBId, roundId } as SetRoundRequest);
        },
    }
});

export const nextRoundStoreKey: InjectionKey<Store<NextRoundStore>> = Symbol();

export function useNextRoundStore(): Store<NextRoundStore> {
    return useStore(nextRoundStoreKey);
}
