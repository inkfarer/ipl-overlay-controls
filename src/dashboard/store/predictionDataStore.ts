import { NodeCGBrowser } from 'nodecg/browser';
import { createStore, Store, useStore } from 'vuex';
import cloneDeep from 'lodash/cloneDeep';
import { InjectionKey } from 'vue';
import { Prediction, PredictionStore } from 'schemas';
import { PredictionStatus } from 'types/enums/predictionStatus';

export const predictionStore = nodecg.Replicant<PredictionStore>('predictionStore');

export const predictionReps = [ predictionStore ];

export interface PredictionDataStore {
    predictionStore: PredictionStore;
}

export const predictionDataStore = createStore<PredictionDataStore>({
    state: {
        predictionStore: null,
    },
    mutations: {
        setState(store, { name, val }: { name: string, val: unknown }): void {
            this.state[name] = cloneDeep(val);
        },
    },
    actions: {
        async lockPrediction(store) {
            return patchPrediction(
                store.state.predictionStore.currentPrediction,
                PredictionStatus.LOCKED,
                [ PredictionStatus.ACTIVE ]);
        },
        async cancelPrediction(store) {
            return patchPrediction(
                store.state.predictionStore.currentPrediction,
                PredictionStatus.CANCELED,
                [ PredictionStatus.ACTIVE, PredictionStatus.LOCKED ]);
        },
        async resolvePrediction(store, { winningOutcomeIndex }: { winningOutcomeIndex: number }) {
            if (winningOutcomeIndex > 1 || winningOutcomeIndex < 0) {
                throw new Error(`Cannot resolve prediction with outcome index ${winningOutcomeIndex}.`);
            } else if (!store.state.predictionStore.currentPrediction?.id) {
                throw new Error('No prediction to resolve.');
            } else if (store.state.predictionStore.currentPrediction.status !== PredictionStatus.LOCKED ) {
                throw new Error('Can only resolve a locked prediction.');
            }

            return nodecg.sendMessage('patchPrediction', {
                id: store.state.predictionStore.currentPrediction.id,
                status: PredictionStatus.RESOLVED,
                winning_outcome_id: store.state.predictionStore.currentPrediction.outcomes[winningOutcomeIndex].id
            });
        },
        async createPrediction(
            store,
            { title, teamAName, teamBName, duration }:
                { title: string, teamAName: string, teamBName: string, duration: number }
        ) {
            if (
                [PredictionStatus.ACTIVE, PredictionStatus.LOCKED]
                    .includes(store.state.predictionStore.currentPrediction?.status as PredictionStatus)
            ) {
                throw new Error('An unresolved prediction already exists.');
            }

            return nodecg.sendMessage('postPrediction', {
                title,
                outcomes: [
                    { title: teamAName },
                    { title: teamBName }
                ],
                prediction_window: duration
            });
        },
        async reconnect() {
            return nodecg.sendMessage('reconnectToRadiaSocket');
        }
    }
});

async function patchPrediction(
    currentPrediction: Prediction,
    status: PredictionStatus,
    allowedForStatuses: PredictionStatus[]
): Promise<void> {
    if (!currentPrediction || !allowedForStatuses.includes(currentPrediction?.status as PredictionStatus)) {
        throw new Error(`Cannot set prediction status at this time. Status must match one of [${allowedForStatuses.join(', ')}]`);
    }

    return nodecg.sendMessage('patchPrediction', {
        id: currentPrediction.id,
        status: status,
    });
}

export const predictionDataStoreKey: InjectionKey<Store<PredictionDataStore>> = Symbol();

export function usePredictionDataStore(): Store<PredictionDataStore> {
    return useStore(predictionDataStoreKey);
}
