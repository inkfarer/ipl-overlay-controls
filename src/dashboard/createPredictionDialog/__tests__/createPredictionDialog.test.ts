import CreatePredictionDialog from '../createPredictionDialog.vue';
import { config, flushPromises, mount } from '@vue/test-utils';
import { createStore } from 'vuex';
import { PredictionDataStore, predictionDataStoreKey } from '../../store/predictionDataStore';
import { PredictionStatus } from 'types/enums/predictionStatus';
import { NextRoundStore, nextRoundStoreKey } from '../../store/nextRoundStore';
import { mockDialog, mockGetDialog } from '../../__mocks__/mockNodecg';
import { closeDialog } from '../../helpers/dialogHelper';
import { PlayType } from 'types/enums/playType';

jest.mock('../../helpers/dialogHelper');

describe('CreatePredictionDialog', () => {
    config.global.stubs = {
        IplInput: true,
        FontAwesomeIcon: true,
        IplErrorDisplay: true,
        IplDialogTitle: true
    };

    const mockCreatePrediction = jest.fn();

    function createNextRoundStore() {
        return createStore<NextRoundStore>({
            state: {
                nextRound: {
                    teamA: { id: '123123', name: 'cool team A', showLogo: true, players: []},
                    teamB: { id: '345345', name: 'cool team B', showLogo: false, players: []},
                    round: { id: '0387', name: 'dope round', type: PlayType.PLAY_ALL },
                    showOnStream: true,
                    games: []
                }
            },
            mutations: {
                setShowOnStream: jest.fn()
            },
            actions: {
                beginNextMatch: jest.fn(),
                setNextRound: jest.fn()
            }
        });
    }

    function createPredictionDataStore() {
        return createStore<PredictionDataStore>({
            state: {
                predictionStore: {
                    status: {
                        socketOpen: true,
                        predictionsEnabled: true
                    },
                    currentPrediction: {
                        id: 'prediction123',
                        broadcasterId: 'ipl',
                        broadcasterName: 'IPL',
                        broadcasterLogin: 'eye pee el',
                        title: 'Who will win?',
                        outcomes: [
                            {
                                id: 'outcome-1',
                                title: 'First Team',
                                users: 5,
                                pointsUsed: 10000,
                                topPredictors: [],
                                color: 'BLUE'
                            },
                            {
                                id: 'outcome-2',
                                title: 'Second Team',
                                users: 1,
                                pointsUsed: 1,
                                topPredictors: [],
                                color: 'PINK'
                            }
                        ],
                        duration: 60,
                        status: PredictionStatus.ACTIVE,
                        creationTime: '2020',
                    }
                }
            },
            actions: {
                createPrediction: mockCreatePrediction,
            }
        });
    }

    it('matches snapshot when existing prediction is locked', () => {
        const predictionDataStore = createPredictionDataStore();
        predictionDataStore.state.predictionStore.currentPrediction.status = PredictionStatus.LOCKED;
        const nextRoundStore = createNextRoundStore();
        const wrapper = mount(CreatePredictionDialog, {
            global: {
                plugins: [[predictionDataStore, predictionDataStoreKey], [nextRoundStore, nextRoundStoreKey]]
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('matches snapshot when existing prediction is active', () => {
        const predictionDataStore = createPredictionDataStore();
        predictionDataStore.state.predictionStore.currentPrediction.status = PredictionStatus.ACTIVE;
        const nextRoundStore = createNextRoundStore();
        const wrapper = mount(CreatePredictionDialog, {
            global: {
                plugins: [[predictionDataStore, predictionDataStoreKey], [nextRoundStore, nextRoundStoreKey]]
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('matches snapshot', () => {
        const predictionDataStore = createPredictionDataStore();
        predictionDataStore.state.predictionStore.currentPrediction.status = PredictionStatus.RESOLVED;
        const nextRoundStore = createNextRoundStore();
        const wrapper = mount(CreatePredictionDialog, {
            global: {
                plugins: [[predictionDataStore, predictionDataStoreKey], [nextRoundStore, nextRoundStoreKey]]
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('disables submit button if any fields are invalid', async () => {
        const predictionDataStore = createPredictionDataStore();
        predictionDataStore.state.predictionStore.currentPrediction.status = PredictionStatus.RESOLVED;
        const nextRoundStore = createNextRoundStore();
        const wrapper = mount(CreatePredictionDialog, {
            global: {
                plugins: [[predictionDataStore, predictionDataStoreKey], [nextRoundStore, nextRoundStoreKey]]
            }
        });

        wrapper.getComponent('[type="number"]').vm.$emit('update:modelValue', 0);
        await wrapper.vm.$nextTick();

        expect(wrapper.getComponent('[data-test="create-prediction-button"]').props().disabled).toEqual(true);
    });

    it('handles creating prediction and closes dialog on create button click', async () => {
        const predictionDataStore = createPredictionDataStore();
        predictionDataStore.state.predictionStore.currentPrediction.status = PredictionStatus.RESOLVED;
        const nextRoundStore = createNextRoundStore();
        const wrapper = mount(CreatePredictionDialog, {
            global: {
                plugins: [[predictionDataStore, predictionDataStoreKey], [nextRoundStore, nextRoundStoreKey]]
            }
        });

        wrapper.getComponent('[type="number"]').vm.$emit('update:modelValue', 128);
        wrapper.getComponent('[data-test="create-prediction-button"]').vm.$emit('click');
        await flushPromises();

        expect(mockCreatePrediction).toHaveBeenCalledWith(expect.any(Object), {
            title: 'Who do you think will win this match?',
            duration: 128,
            teamAName: 'cool team A',
            teamBName: 'cool team B'
        });
        expect(mockGetDialog).toHaveBeenCalledWith('createPredictionDialog');
        expect(mockDialog.close).toHaveBeenCalledTimes(1);
    });

    it('resets data to defaults on reset button click', async () => {
        const predictionDataStore = createPredictionDataStore();
        predictionDataStore.state.predictionStore.currentPrediction.status = PredictionStatus.RESOLVED;
        const nextRoundStore = createNextRoundStore();
        const wrapper = mount(CreatePredictionDialog, {
            global: {
                plugins: [[predictionDataStore, predictionDataStoreKey], [nextRoundStore, nextRoundStoreKey]]
            }
        });

        wrapper.getComponent('[type="number"]').vm.$emit('update:modelValue', 128);
        wrapper.getComponent('[name="title"]').vm.$emit('update:modelValue', 'Who will win?');
        wrapper.getComponent('[name="team-a-name"]').vm.$emit('update:modelValue', 'Team One!!!');
        wrapper.getComponent('[name="team-b-name"]').vm.$emit('update:modelValue', 'Team Two?!?!?!?');
        wrapper.getComponent('[data-test="reset-inputs-button"]').vm.$emit('click');
        await wrapper.vm.$nextTick();

        expect(wrapper.getComponent('[type="number"]').attributes().modelvalue).toEqual('120');
        expect(wrapper.getComponent('[name="title"]').attributes().modelvalue).toEqual('Who do you think will win this match?');
        expect(wrapper.getComponent('[name="team-a-name"]').attributes().modelvalue).toEqual('cool team A');
        expect(wrapper.getComponent('[name="team-b-name"]').attributes().modelvalue).toEqual('cool team B');
    });

    it('closes dialog on dialog title close event', () => {
        const predictionDataStore = createPredictionDataStore();
        const nextRoundStore = createNextRoundStore();
        const wrapper = mount(CreatePredictionDialog, {
            global: {
                plugins: [[predictionDataStore, predictionDataStoreKey], [nextRoundStore, nextRoundStoreKey]]
            }
        });

        wrapper.getComponent('ipl-dialog-title-stub').vm.$emit('close');

        expect(closeDialog).toHaveBeenCalledWith('createPredictionDialog');
    });
});
