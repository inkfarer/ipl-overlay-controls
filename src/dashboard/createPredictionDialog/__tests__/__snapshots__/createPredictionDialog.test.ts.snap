// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatePredictionDialog matches snapshot 1`] = `
"<ipl-dialog-title-stub title=\\"Create Prediction\\"></ipl-dialog-title-stub>
<ipl-error-display-stub class=\\"m-t-8\\"></ipl-error-display-stub>
<div class=\\"ipl-space color-dark m-t-8\\" data-v-11398c0e=\\"\\">
  <div class=\\"layout horizontal\\">
    <ipl-input-stub name=\\"title\\" label=\\"Title\\" modelvalue=\\"Who do you think will win this match?\\" type=\\"text\\" validator=\\"[object Object]\\" centered=\\"false\\" disabled=\\"false\\" class=\\"title-input\\"></ipl-input-stub>
    <ipl-input-stub name=\\"duration\\" label=\\"Duration\\" modelvalue=\\"120\\" type=\\"number\\" validator=\\"[object Object]\\" centered=\\"false\\" disabled=\\"false\\" extra=\\"seconds\\" class=\\"m-l-6 duration-input\\"></ipl-input-stub>
  </div>
  <div class=\\"layout horizontal m-t-4\\">
    <ipl-input-stub name=\\"team-a-name\\" label=\\"Team A\\" modelvalue=\\"cool team A\\" type=\\"text\\" validator=\\"[object Object]\\" centered=\\"false\\" disabled=\\"false\\" class=\\"max-width\\"></ipl-input-stub>
    <ipl-input-stub name=\\"team-b-name\\" label=\\"Team B\\" modelvalue=\\"cool team B\\" type=\\"text\\" validator=\\"[object Object]\\" centered=\\"false\\" disabled=\\"false\\" class=\\"max-width m-l-6\\"></ipl-input-stub>
  </div><a class=\\"ipl-button m-t-8\\" href=\\"javascript:void(0);\\" style=\\"background-color: rgb(0, 166, 81); color: white;\\" data-test=\\"create-prediction-button\\"><span class=\\"label\\">Create</span></a>
</div>"
`;

exports[`CreatePredictionDialog matches snapshot when existing prediction is active 1`] = `
"<ipl-dialog-title-stub title=\\"Create Prediction\\"></ipl-dialog-title-stub>
<ipl-error-display-stub class=\\"m-t-8\\"></ipl-error-display-stub>
<div class=\\"ipl-message__wrapper layout horizontal center-vertical ipl-message__type-warning m-t-8\\" data-v-13b37b8d=\\"\\">
  <font-awesome-icon-stub icon=\\"exclamation-triangle\\" border=\\"false\\" fixedwidth=\\"false\\" listitem=\\"false\\" pulse=\\"false\\" swapopacity=\\"false\\" spin=\\"false\\" symbol=\\"false\\" inverse=\\"false\\" class=\\"icon\\" data-v-13b37b8d=\\"\\"></font-awesome-icon-stub>
  <div class=\\"message-content\\" data-v-13b37b8d=\\"\\">
    <!----> The active prediction has not been resolved!
  </div>
</div>"
`;

exports[`CreatePredictionDialog matches snapshot when existing prediction is locked 1`] = `
"<ipl-dialog-title-stub title=\\"Create Prediction\\"></ipl-dialog-title-stub>
<ipl-error-display-stub class=\\"m-t-8\\"></ipl-error-display-stub>
<div class=\\"ipl-message__wrapper layout horizontal center-vertical ipl-message__type-warning m-t-8\\" data-v-13b37b8d=\\"\\">
  <font-awesome-icon-stub icon=\\"exclamation-triangle\\" border=\\"false\\" fixedwidth=\\"false\\" listitem=\\"false\\" pulse=\\"false\\" swapopacity=\\"false\\" spin=\\"false\\" symbol=\\"false\\" inverse=\\"false\\" class=\\"icon\\" data-v-13b37b8d=\\"\\"></font-awesome-icon-stub>
  <div class=\\"message-content\\" data-v-13b37b8d=\\"\\">
    <!----> The active prediction has not been resolved!
  </div>
</div>"
`;
